trie/branchNode.go

```diff
@@ -653,36 +717,74 @@ func (bn *branchNode) delete(data []core.TrieData, db common.TrieStorageInteract
 
 	numChildren, pos := getChildPosition(bn)
 	if numChildren == 0 {
-		return true, nil, modifiedHashes, nil
+		return true, nil, modifiedHashes
 	}
 	if numChildren == 1 {
 		err = resolveIfCollapsed(bn, byte(pos), db)
 		if err != nil {
-			return false, nil, emptyHashes, err
+			goRoutinesManager.SetError(err)
+			return false, nil, [][]byte{}
 		}
 
 		err = resolveIfCollapsed(bn.children[pos], byte(pos), db)
 		if err != nil {
-			return false, nil, emptyHashes, err
+			goRoutinesManager.SetError(err)
+			return false, nil, [][]byte{}
 		}
 
 		var newChildHash bool
 		newNode, newChildHash, err := bn.children[pos].reduceNode(pos)
 		if err != nil {
-			return false, nil, emptyHashes, err
+			goRoutinesManager.SetError(err)
+			return false, nil, [][]byte{}
 		}
 
 		if newChildHash && !bn.children[pos].isDirty() {
 			modifiedHashes = append(modifiedHashes, bn.children[pos].getHash())
 		}
 
-		return true, newNode, modifiedHashes, nil
+		return true, newNode, modifiedHashes
 	}
 
-	return true, bn, modifiedHashes, nil
+	return true, bn, modifiedHashes
+}
+
+func (bn *branchNode) deleteChild(
+	dataForRemoval []core.TrieData,
+	childPos int,
+	hasBeenModified *atomic.Flag,
+	goRoutinesManager common.TrieGoroutinesManager,
+	db common.TrieStorageInteractor,
+) [][]byte {
+	err := resolveIfCollapsed(bn, byte(childPos), db)
+	if err != nil {
+		goRoutinesManager.SetError(err)
+		return [][]byte{}
+	}
+
+	if bn.children[childPos] == nil {
+		return [][]byte{}
+	}
+
+	dirty, newNode, oldHashes := bn.children[childPos].delete(dataForRemoval, goRoutinesManager, db)
+	if !goRoutinesManager.ShouldContinueProcessing() || !dirty {
+		return oldHashes
+	}
+
+	err = bn.setNewChild(byte(childPos), newNode)
+	if err != nil {
+		goRoutinesManager.SetError(err)
+		return [][]byte{}
+	}
+
+	hasBeenModified.SetValue(true)
+	return oldHashes
 }
 
 func (bn *branchNode) setNewChild(childPos byte, newNode node) error {
+	bn.mutex.Lock()
+	defer bn.mutex.Unlock()
```

Comment:
for which variable(s) is the lock? i think we need to add lock also in other places

Reply:
These are for dirty, hash, version. Indeed, this will need to be added in other places. Will do that in another PR, in which I will also write some tests for this cases.